Foundation/NSURL.swift

```diff
@@ -638,13 +665,462 @@ open class NSURL : NSObject, NSSecureCoding, NSCopying {
     override open var _cfTypeID: CFTypeID {
         return CFURLGetTypeID()
     }
+    
+    private func _initResourceInfo() {
+        let pInfo = UnsafeMutablePointer<NSMutableArray>.allocate(capacity: 1)
+        pInfo.initialize(to: NSMutableArray())
+        _CFSwiftRetain(UnsafeMutableRawPointer(pInfo)) // Balanced with resourceInfo release in _CFURLDeallocate()
+        pInfo.pointee.add(NSMutableDictionary())
+        pInfo.pointee.add(NSLock())
+        __CFURLSetResourceInfoPtr(_cfObject, UnsafeMutableRawPointer(pInfo))
+    }
+    
+    private static let resourceInfoInitLock = NSLock()
+    
+    private func _accessResourceDictionary<T>(_ handler: (_ resourcesCache: NSMutableDictionary) -> T) -> T? {
+        // Resource values are only accessible for file urls.
+        guard _CFURLIsFileURL(_cfObject) else {
+            return nil
+        }
+        
+        // We don't have a lock for the instance's resourceInfo yet, and we don't want to initialize it twice!
+        NSURL.resourceInfoInitLock.lock()
+        if __CFURLResourceInfoPtr(_cfObject) == nil {
+            _initResourceInfo()
+        }
+        NSURL.resourceInfoInitLock.unlock()
+        let resourceInfoPtr = __CFURLResourceInfoPtr(_cfObject)!.assumingMemoryBound(to: NSMutableArray.self)
+        let resourcesCache = resourceInfoPtr.pointee.object(at: 0) as! NSMutableDictionary
+        let lock = resourceInfoPtr.pointee.object(at: 1) as! NSLock
+        
+        lock.lock()
+        defer {
+            lock.unlock()
+        }
+        return handler(resourcesCache)
+    }
+    
+    open func removeAllCachedResourceValues() {
+        _accessResourceDictionary { (resourcesCache) in
+            resourcesCache.removeAllObjects()
+        }
+    }
+    
+    open func removeCachedResourceValue(forKey key: URLResourceKey) {
+        _accessResourceDictionary { (resourcesCache) in
+            resourcesCache.removeObject(forKey: key.rawValue)
+        }
+    }
+    
+    open func resourceValues(forKeys keys: [URLResourceKey]) throws -> [URLResourceKey : Any] {
+        
+        func hasFlag(path: String, flag: Int32) throws -> Bool {
+            #if canImport(Darwin) || os(FreeBSD)
+            let stat = try FileManager.default._lstatFile(atPath: path)
+            return stat.st_flags & UInt32(flag) != 0
+            #elseif canImport(Glibc)
+            let fh = try FileHandle(forUpdating: self._swiftObject)
+            var attrs: Int32 = 0
+            if ioctl(fh.fileDescriptor, FS_IOC_GETFLAGS, &attrs) != 0 {
+                throw _NSErrorWithErrno(errno, reading: true, path: path)
+            }
+            return attrs & flag != 0
+            #elseif os(Windows)
```

Comment:
For now it might be bets to change this whole section  to 
```
#else
#error("Requires a platform-specific implementation")
#endif
```

As it will notify the platform maintainers for Windows/Android etc to add the appropriate support.